src/OpenSage.Game/Logic/Object/Body/ActiveBody.cs

```diff
@@ -84,25 +89,28 @@ public override void DoDamage(DamageType damageType, Fix64 amount, DeathType dea
         ValidateArmorAndDamageFX();
 
         // Actual amount of damage depends on armor.
-        var damagePercent = _currentArmor?.GetDamagePercent(damageType) ?? new Percentage(1.0f);
-        var actualDamage = amount * (Fix64)(float)damagePercent;
+        var damagePercent = _currentArmor?.GetDamagePercent(damageInfo.Request.DamageType) ?? new Percentage(1.0f);
+        var actualDamage = (Fix64)damageInfo.Request.DamageToDeal * (Fix64)(float)damagePercent;
 
-        var takingDamage = damageType is not DamageType.Healing;
+        var takingDamage = damageInfo.Request.DamageType is not DamageType.Healing;
 
         var newHealth = takingDamage ? Health - actualDamage : Health + actualDamage;
 
         SetHealth(newHealth);
 
-        var damageRequest = new DamageDataRequest(damageDealer?.ID ?? 0, damageType, deathType, (float)amount, damageDealer?.Definition.Name ?? string.Empty);
-        var damageResult = new DamageDataResult((float)actualDamage, _lastHealthBeforeDamage - _currentHealth);
+        var damager = _context.GameLogic.GetObjectById(damageInfo.Request.DamageDealer);
+        damageInfo.Request.AttackerName = damager?.Definition.Name ?? string.Empty;
```

Comment:
Would it be better to set this on construction of the `DamageInfo` object? Or is this done here specifically to avoid setting this when the target's health is zero?

Just thinking if we don't need to set this here, this doesn't need to be a ref parameter and we don't need to expose the field to be written publicly when it _shouldn't_(?) change

Reply:
I was more-or-less copying the [original code](https://github.com/electronicarts/CnC_Generals_Zero_Hour/blob/main/GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Body/ActiveBody.cpp#L370) here, although I'm sure it can be cleaned up a bit when we come to port `ActiveBody.cpp`.

I'm actually not sure yet whether it needs to be a `ref` parameter - I was copying what the original does, but I don't know if it's ever necessary for calling code to access the modified "request". I suggest we leave it as it is for now until we port this code properly.